File: lib/display/display.dart
import 'dart:io';

import 'package:colorgram/colorgram.dart';
import 'package:http/http.dart' as http;
import 'package:image/image.dart' as img;
import 'package:orbit/api.dart';

/// Abstract class for the multiples way of displaying the infos from the api (in a terminal, on a led strip...)
abstract class Display {
  final LastFM api;

  /// The number of colors extracted from the album cover.
  final int colorNUmber = 3;

  Display(this.api);

  Future<List<CgColor>?> getColorPalette(String imageUrl) async {
    try {
      final response = await http.get(Uri.parse(imageUrl));

      if (response.statusCode == 200) {
        // Décoder l'image en mémoire
        img.Image? image = img.decodeImage(response.bodyBytes);

        if (image != null) {
          // Solution temporaire: écrire dans un fichier mémoire
          final tempFile = File('temp_image.jpg');
          await tempFile.writeAsBytes(img.encodeJpg(image));

          List<CgColor> colors = extractColor(tempFile, colorNUmber);

          await tempFile.delete();

          return colors;
        }
      }
    } catch (e) {
      print('Erreur: $e');
      return null;
    }
    return null;
  }

  Future<void> display();
}

File: lib/display/display.py
import os
from abc import ABC, abstractmethod

import requests
from Pylette import extract_colors, Palette

from lib.api import LastFM


class Display(ABC):
    api: LastFM

    # The number of colors extracted from the album cover.
    colorNumber:int = 12

    def __init__(self, api: LastFM):
        self.api = api

    def getColorPalette(self, imageUrl:str, recur:bool = False) -> Palette:
        img = requests.get(imageUrl).content
        tempFile = open("temp.jpg", "wb")

        tempFile.write(img)
        tempFile.close()

        try:
            palette = extract_colors(image='temp.jpg', palette_size=self.colorNumber)
        except Exception as e:
            if recur:
                # get placeholder image by default, might change that in config file tho
                palette = extract_colors(image='assets/placeholder.jpeg', palette_size=self.colorNumber)
                #raise Exception(f"while getting color palette for image '{imageUrl}' the following error occured: {e}")
            else :
                # retry one time
                palette = self.getColorPalette(imageUrl, recur=True)

        os.remove("temp.jpg")
        return palette

    @abstractmethod
    def display(self): pass


